backend-public/contracts/ryder-handles-controller.clar

```diff
@@ -0,0 +1,195 @@
+    (map-set renewal-signatures approval-signature true)
+    (try! (pay-fees-from-escrow (get price preorder) namespace))
+    (try! (stx-transfer? u1 tx-sender (as-contract tx-sender)))
+    (try! (as-contract (contract-call? .community-handles name-register namespace name owner zonefile-hash)))
+    (ok true)))
+
+
+;; renew a name
+;; a new signature is required for each renewal
+;; @param name; the name to renew
+;; @param salt; the salt used by the approver
+;; @param approval-signature; signed hash by the approver
+;; @param new-owner;
+;; @param zonefile-hash;
+(define-public (name-renewal (namespace (buff 20))
+                             (name (buff 48))
+                             (salt (buff 20))
+                             (approval-signature (buff 65))
+                             (new-owner (optional principal))
+                             (zonefile-hash (optional (buff 20))))
+  (let ((price (var-get price-in-ustx))
+        (owner tx-sender)
+        (hash (sha256 (concat (concat (concat name 0x2e) namespace) salt))))
+      ;; signature must be correct
+      (asserts! (secp256k1-verify hash approval-signature (var-get approval-pubkey)) err-not-authorized)
+      ;; signature must be unused
+      (asserts! (is-none (map-get? renewal-signatures approval-signature)) err-signature-already-used) 
+      (map-set renewal-signatures approval-signature true)
+      (try! (pay-fees price (some namespace)))
+      (try! (contract-call? .community-handles name-renewal namespace name new-owner zonefile-hash))
+      (ok true)))
+
+(define-private (pay-fees (price uint) (namespace-or-escrow (optional (buff 20))))
+  (let ((amount-controller-admin (/ (* price u70) u100))
+        (amount-community (- price amount-controller-admin))
+        (controller-admin (var-get contract-owner))        
+        (community-treasury (match namespace-or-escrow
+                              namespace (default-to controller-admin (map-get? community-treasuries namespace))
+                              (as-contract tx-sender))))
+    (and (> amount-controller-admin u0)
+        (try! (stx-transfer? amount-controller-admin tx-sender controller-admin)))
+    (and (> amount-community u0)
+        (try! (stx-transfer? amount-community tx-sender community-treasury)))
+    (ok true)))
+
+(define-private (pay-fees-from-escrow (price uint) (namespace (buff 20)))
+    (let ((amount-controller-admin (/ (* price u70) u100))
+          (amount-community (- price amount-controller-admin))
+          (community-treasury (default-to (var-get contract-owner) (map-get? community-treasuries namespace))))
+        (and (> amount-community u0)
+            (try! (as-contract (stx-transfer? amount-community tx-sender community-treasury))))
+        (ok true)))
+
+;;
+;; admin functions
+;;
+(define-public (set-price (amount-in-ustx uint))
+    (begin
+        (try! (is-contract-owner))
+        (var-set price-in-ustx amount-in-ustx)
+        (ok true)))
```

Comment:
Should price also be settable per namespace? Since the controller can handle multiple.

Reply:
If it is more clear what each community wants each namespace should eventually have their own controller.